### PR TITLE
Cleanup memory pool code add some tests

### PR DESCRIPTION
**Pull request by pakrym**
This PR adds more descriptive error messages to existing cases and adds two more checks in debug mode:

1. Double dispose of pool
2. Disposing pool with active blocks.

It also adds one check in release mode:

1. Renting from a disposed pool. Current behaviour there is highly unpredictable and hard to correctly handle.

And removes one check in release mode:

1. Disposing (returning) block to a disposed pool.

This PR doesn't address the unpinning while in use issue raised by @benaadams  but I don't want to mix up cleanup/diagnostics and functional fixes.

### CODE REVIEW

File: shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/MemoryPoolBlock.Debug.cs [hunk 1]
```diff
@@ -24,100 +34,105 @@ internal sealed class MemoryPoolBlock : MemoryManager<byte>
...
             }
 
-            Pool.Return(this);
+            _pool.Return(this);
+
+            // Let finalizer to run in _pool.Return so we don't crash
+            // process with Debug.Assert
+            if (!_slab.IsActive)
+            {
+                MemoryPoolThrowHelper.ThrowInvalidOperationException_BlockReturnedToDisposedPool();
```
Comment (benaadams) [inline shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/MemoryPoolBlock.Debug.cs]: Kill this exception in Release mode? Its still there...

File: shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/MemoryPoolBlock.Debug.cs [hunk 2]
```diff
@@ -24,100 +34,118 @@ internal sealed class MemoryPoolBlock : MemoryManager<byte>
...
-        public MemoryPoolSlab Slab { get; }
-
         public override Memory<byte> Memory
         {
             get
             {
-                if (!Slab.IsActive) MemoryPoolThrowHelper.ThrowObjectDisposedException(MemoryPoolThrowHelper.ExceptionArgument.MemoryPoolBlock);
+                if (!_slab.IsActive)
+                {
+                    MemoryPoolThrowHelper.ThrowObjectDisposedException(MemoryPoolThrowHelper.ExceptionArgument.MemoryPoolBlock);
```
Comment (halter73) [inline shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/MemoryPoolBlock.Debug.cs]: Would it be better to have an _isDisposed field in MemoryPoolBlock itself that was set in Dispose and unset in Lease? Then there'd be a different exception for using the block after it's returned to the pool and using the block before it's returned to the pool but after the pool is disposed.

Comment (pakrym) [reply shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/MemoryPoolBlock.Debug.cs]: Thought about it too, will do.

File: shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/MemoryPoolSlab.cs [hunk 3]
```diff
@@ -38,13 +34,11 @@ public MemoryPoolSlab(byte[] data)
         /// collected and the slab is no longer references the slab will be garbage collected and the memory unpinned will
         /// be unpinned by the slab's Dispose.
         /// </summary>
-        public bool IsActive => _isActive;
-
-        public IntPtr NativePointer => _nativePointer;
+        public bool IsActive { get; private set; }
```
Comment (halter73) [inline shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/MemoryPoolSlab.cs]: Is this ever different than `public bool IsActive => !_isDisposed;`

Comment (pakrym) [reply shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/MemoryPoolSlab.cs]: Nope.

File: shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/SlabMemoryPool.cs [hunk 4]
```diff
@@ -119,6 +127,7 @@ private MemoryPoolBlock AllocateSlab()
                 block.IsLeased = true;
 #endif
                 Return(block);
+                blocksAllocated++;
```
[the remark this answers is not in the document]
Comment (pakrym) [reply shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/SlabMemoryPool.cs]: We create additional segment after the loop. It takes the last `_blockSize` of memory.

**Comment by pakrym** [comment]
:up::date:

File: shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/SlabMemoryPool.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (halter73) [reply shared/Microsoft.Extensions.Buffers.MemoryPool.Sources/SlabMemoryPool.cs]: Ahhhh. I think it's technically right then but still a little hard to follow. I think it would be a lot easier to grok (if only because less bitwise operations) if written as follows:

```c#
// ...

var maxPenultimateOffset = _slabLength - (_blockSize * 2);
var offset = firstOffset;
for (;
    offset <= maxPenultimateOffset;
    offset += _blockSize)
{
    var block = new MemoryPoolBlock(
        this,
        slab,
        offset,
        _blockSize);
    
    // ...
```
